File: Session17/416. Partition Equal Subset Sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum=0;
        
        for(int s:nums)
        {
            sum+=s;
        }
        
        if(sum%2!=0)
        {
            return false;
        }
        
        sum=sum/2;
        int n=nums.length;
        boolean dp[]=new boolean[sum+1];
        
        dp[0]=true;
        
        for(int i=0;i<n;i++)
        {
            for(int j=sum;j>=nums[i];j--)
            {
                dp[j]=dp[j]||dp[j-nums[i]];
            }
        }
        
        return dp[sum];
        
        
    }
}

File: Session17/416. Partition Equal Subset Sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        sum=0
        
        for num in nums:
            sum+=num
        
        if sum%2!=0:
            return False
        
        sum=int(sum/2)
        
        
        dp=[False for i in range(sum+1)]
        
        dp[0]=True
        
        for i in range(len(nums)):
            for j in range(sum,-1,-1):
                if j>=nums[i]:
                    dp[j] = dp[j] or dp[j-nums[i]]
        
        return dp[sum]
    
    